src/dataflow/src/render/reduce.rs

```diff
@@ -1095,80 +1227,94 @@ where
-                    Datum::Float32(f) => ((f64::from(*f) * float_scale) as i128, 1),
-                    Datum::Float64(f) => ((*f * float_scale) as i128, 1),
-                    Datum::Decimal(d) => (d.as_i128(), 1),
-                    Datum::Null => (0, 0),
+                    Datum::Int32(i) => Accumulator::SimpleNumber {
+                        sum: i128::from(i),
+                        count: 1,
+                    },
+                    Datum::Int64(i) => Accumulator::SimpleNumber {
+                        sum: i128::from(i),
+                        count: 1,
+                    },
+                    Datum::Float32(f) => Accumulator::SimpleNumber {
+                        sum: (f64::from(*f) * float_scale) as i128,
+                        count: 1,
+                    },
+                    Datum::Float64(f) => Accumulator::SimpleNumber {
+                        sum: (*f * float_scale) as i128,
+                        count: 1,
+                    },
+                    Datum::Decimal(d) => Accumulator::SimpleNumber {
+                        sum: d.as_i128(),
+                        count: 1,
+                    },
+                    Datum::Null => Accumulator::SimpleNumber { sum: 0, count: 0 },
                     x => panic!("Accumulating non-integer data: {:?}", x),
                 }
             }
         }
     };
 
     let mut to_aggregate = Vec::new();
-    let diffs_len = num_accumulable * 3;
     // First, collect all non-distinct aggregations in one pass.
-    let easy_cases = collection
-        .inner
-        .map(|(d, t, r)| (d, t, r as i128))
-        .as_collection()
-        .explode({
-            move |(key, row)| {
-                let mut diffs = vec![0i128; diffs_len];
-                // Try to unpack only the datums we need. Unfortunately, since we
-                // can't random access into a Row, we have to iterate through one by one.
-                // TODO: Even though we don't have random access, we could still avoid unpacking
-                // everything that we don't care about, and it might be worth it to extend the
-                // Row API to do that.
-                let mut row_iter = row.iter().enumerate();
-                for (accumulable_index, datum_index, aggr) in simple_aggrs.iter() {
-                    let mut datum = row_iter.next().unwrap();
-                    while datum_index != &datum.0 {
-                        datum = row_iter.next().unwrap();
-                    }
-                    let datum = datum.1;
-                    let (agg1, agg2) = datum_aggr_values(datum, &aggr.func);
-                    diffs[3 * accumulable_index] = 1i128;
-                    diffs[3 * accumulable_index + 1] = agg1;
-                    diffs[3 * accumulable_index + 2] = agg2;
+    let easy_cases = collection.inner.as_collection().explode({
+        move |(key, row)| {
+            let mut diffs = vec![Accumulator::default(); num_accumulable];
```

Comment:
so I don't think its legit to do this here, because the ith accumulator may not match the default type (I'm pretty sure thats why the test is panicking)

Reply:
the intent here is if you did something like
```
sum(x), sum(distinct y), sum(z), sum(distinct a)
```

you would get 3 sets of diffs per key

one containing `[accumulator for sum(x), 0, accumulator for sum(z), 0]` i.e. one for all the non-distinct or easy cases, and then one each for the distinct cases e.g.

`[0, accumulator for sum(distinct y), 0, 0]` and `[0, 0, 0 accumulator for sum(distinct a)]` and then all 3 (or really n_distinct aggregations + 1) diffs would get merged together. i hope that makes sense

Reply:
I think you should be able to use `vec![]` as the default. The `Vec` implementation of `Difference` treats absent coordinates as zero, without needing a zero to populate them. I think. It's meant to, at least.

Reply:
Actually, I take that back I have no clue what is going on here, but ... if the panicking is happening because `default()` has variant `Count` which cannot be added to anything else, starting with an empty vector should sort it out. @ruchirK has various comments about the distinct vs non-distinct stuff, and tbh I forgot about that and don't know what is up. :D

Reply:
but then you would need to be more careful about how you populate something at an arbitrary index right? imo plus one could just check either if the rhs or lhs is_zero and in that case, forward the opposite value

Reply:
I was just explaining why its set up the way that it is -- specifically that we need to see a specific number of accumulators in each vector, and each accumulator needs to be in the right position. I _think_ based on the last time I tried to do this starting with an empty vector makes this slightly harder because then if you want to put a value at index say 3, you have to remember to also fill indexes 0, 1, and 2 with something

Reply:
The thing I don't know tbh is whether the correct action is an interleaving of the results or a sum. Probably an interleaving, right?

Reply:
Or, wait, your example suggests that it wouldn't be an interleaving, and the contributions would each have explicit zeros in them. But probably needs careful logic in either case. Seems like you should be able to look at the contributions, and at most one should be non-zero and you should select it, right? Or harden `plus_equals` at the probably negligible cost of some branch.

Reply:
Sorry, the "empty vector" thing was a red herring probably. I was explaining that `vec![]` is a valid argument to `plus_equals` and is treated as having empty values. Doing the vector-wise `plus_equals` could work out, except that it would inherit some zero values from the accumulations (idk where they get installed). Will stop typing without reading the surrounding code. :)

Reply:
I think this is the simple diff that fixes the test failure
```
altaria-2:materialize Test$ git diff
diff --git a/src/dataflow/src/render/reduce.rs b/src/dataflow/src/render/reduce.rs
index b1a1cbb38..52fb2853d 100644
--- a/src/dataflow/src/render/reduce.rs
+++ b/src/dataflow/src/render/reduce.rs
@@ -1123,6 +1123,15 @@ impl Semigroup for Accumulator {
     }

     fn plus_equals(&mut self, other: &Accumulator) {
+        if other.is_zero() {
+            return;
+        }
+
+        if self.is_zero() {
+            *self = other.clone();
+            return;
+        }
+
         match (self, other) {
             (
                 Accumulator::Bool {
```

and I think we actually don't even need this because we know exactly what accumulators we are going to produce and in what order, so we could instantiate a vec full of zero instances of all the accumulators and then clone that whenever we want a new "blank" diff vec if that makes sense

Reply:
its dinnertime here but i can type that up pretty easily

Reply:
As promised, I think this is the other approach one could take
```
altaria-2:materialize Test$ git diff
diff --git a/src/dataflow/src/render/reduce.rs b/src/dataflow/src/render/reduce.rs
index b1a1cbb38..eef0a105d 100644
--- a/src/dataflow/src/render/reduce.rs
+++ b/src/dataflow/src/render/reduce.rs
@@ -1210,7 +1210,7 @@ fn build_accumulable<G>(
         full_aggrs,
         simple_aggrs,
         distinct_aggrs,
-        num_accumulable,
+        num_accumulable: _,
     }: AccumulablePlan,
     prepend_key: bool,
 ) -> Arrangement<G, Row>
@@ -1228,6 +1228,22 @@ where
     // reduced if we want to specialize for the aggregations.

     let float_scale = f64::from(1 << 24);
+    // Instantiate a default vector for diffs with the correct types at each position.
+    let diffs: Vec<_> = full_aggrs
+        .iter()
+        .map(|f| match f.func {
+            AggregateFunc::Count | AggregateFunc::Dummy => Accumulator::Count {
+                non_nulls: 0,
+                total: 0,
+            },
+            AggregateFunc::Any | AggregateFunc::All => Accumulator::Bool {
+                trues: 0,
+                falses: 0,
+                total: 0,
+            },
+            _ => Accumulator::SimpleNumber { sum: 0, count: 0 },
+        })
+        .collect();

     // Two aggregation-specific values for each aggregation.
     let datum_to_accumulator = move |datum: Datum, aggr: &AggregateFunc| {
@@ -1296,8 +1312,9 @@ where
     let mut to_aggregate = Vec::new();
     // First, collect all non-distinct aggregations in one pass.
     let easy_cases = collection.inner.as_collection().explode({
+        let diffs = diffs.clone();
         move |(key, row)| {
-            let mut diffs = vec![Accumulator::default(); num_accumulable];
+            let mut diffs = diffs.clone();
             // Try to unpack only the datums we need. Unfortunately, since we
             // can't random access into a Row, we have to iterate through one by one.
             // TODO: Even though we don't have random access, we could still avoid unpacking
@@ -1320,6 +1337,7 @@ where
     // Next, collect all aggregations that require distinctness.
     for (accumulable_index, datum_index, aggr) in distinct_aggrs.into_iter() {
         let mut packer = RowPacker::new();
+        let diffs = diffs.clone();
         let collection = collection
             .map(move |(key, row)| {
                 let value = row.iter().nth(datum_index).unwrap();
@@ -1332,7 +1350,7 @@ where
             .explode({
                 move |(key, row)| {
                     let datum = row.iter().next().unwrap();
-                    let mut diffs = vec![Accumulator::default(); num_accumulable];
+                    let mut diffs = diffs.clone();
                     diffs[accumulable_index] = datum_to_accumulator(datum, &aggr.func);
                     Some((key, diffs))
                 }
```

which doesn't introduce any additional branches into `plus_equals`. Both of these diffs independently pass the sqllogictest that is currently failing in this branch.

Reply:
I like the look of that second diff. Done!